x-pack/solutions/security/plugins/security_solution/docs/adrs/threat_hunting/0002_timeline_ui_batching_strategy.md

```diff
@@ -0,0 +1,172 @@
+
+- **dataViewId**: Currently selected dataView Id in Timeline.
+- **Timerange**: 
+  - **startDate**
+  - **endDate**
+  - **timeRangeKind**
+- **eqlOptions**
+- **indexNames**: List of indices in Timeline.
+- **id**: Timeline Id
+- **fields**:  List of requested fields
+- **filterQuery**:  combined query consisting: 
+  * **KQL/Lucene Query**
+  * **Filters**
+  * **Data Providers**
+- **runtimeMappings**: List of runtime mappings that need to be fetched
+- **language**: KQL | Lucene
+- **limit**: Sample Size or Batch Size
+- **sort**
+
+## Batch fetch/re-fetch Strategy
+Batch fetch/re-fetch strategy depends on the scenario or the action the user is taking. This section will lay out all the scenarios in which a batch should be re-fetched and its corresponding strategy.
+
+### New Batch Requested
+When a user requests a new batch after clicking on `Load more` as shown below, a new batch will be fetched. For example, if user is on Batch 0 and clicks on Load more, a new batch will be fetched with below strategy :
+
+![](./images/load_more_timeline.png)
+
+```js
+{
+  /* Fetch Batch 1 */
+  activePage: 1, /* current batch */
+  querySize: 500 /* batch/sample size */
+}
+```
+
+Consequently, if a user requests a new batch again ( for the 3rd time ), Batch 2 will be fetched with below strategy.
+
+```js
+{
+	/* Fetch Batch 2 */
+	activePage: 2,
+	querySize: 500 /* batch/sample size */
+}
+```
+ 
+### Refresh Button
+
+When a user clicks on the Refresh button in the query tab, it should explicitly re-fetch new data ir-respective of the fact if any batch parameters have been changed or not.
+
+#### When user is on the first (0th) batch
+When user is on the first batch i.e. Batch 0, it will be re-fetched with the below strategy.
+
+  ```js
+  {
+    /* Fetch Batch 0 */
+    activePage: 0,
+    querySize: 500 /* batch/sample size */
+  }
+  ```
+
+#### When user has already fetched multiple batches
+Let’s say the user has already fetched 3 batches and currently the user is on Batch 2. Clicking on Refetch will reset the result set to 0th batch:
+
+
+```js
+{
+	/* Fetch Batch 0 */
+	activePage: 0,
+	querySize: 500 /* batch/sample size */
+}
+```
+
+### Adding a new column
+Adding a new column should not affect the result set that is currently being displayed in front of the user. We should fetch a complete result set ir-respective of the batch user is on.
+
+#### When user is on the first(0th) batch
+When user is on the first batch i.e. Batch 0, it will be re-fetched with the below strategy.
+
+```js
+{
+	/* Fetch Batch 0 */
+	activePage: 0,
+	querySize: 500 /* batch/sample size */
+}
+```
+
+#### When user has already fetched multiple batches
+
+When user is on the 4th batch i.e. Batch 3, it will be re-fetched with the below strategy. 
+
+```js
+{
+  /* Fetch Batch 0 */
+  activePage: 0,
```

Comment:
is the `activePage` 0 here? Or should it be 4?

Reply:
No, it should `0` since we want to fetch all the data for 4 batches cumulatively.

It is will be considered 0th batch of size 2000, and then user can keep fetching further batches of size 500, for example.

Reply:
ok, it does, thanks for clarifying! Doesn't the UI stay on the table/datagrid pagination page the user is currently on though? I wonder if that would be worth mentioning. In a few places in this document we mention what happens or does not happen on the UI side, maybe here we can explain the expected behavior also?

Reply:
>ok, it does, thanks for clarifying! Doesn't the UI stay on the table/datagrid pagination page the user is currently on though?
 
It does.

>I wonder if that would be worth mentioning. In a few places in this document we mention what happens or does not happen on the UI side, maybe here we can explain the expected behavior also?

I think i agree.. Let me add a section on that.

Reply:
It is a bit deceiving, because yes, while the query is right, the actual active page in the UI doesn't change. 🤔

Reply:
just a suggestion that could be applied to each section: have 2 bullet points:
- expected server side call logic
- expected UI behavior

Reply:
I agree but i will re-word it i a little.

- expected fetch/re-fetch logic
- expected UI behavior

Reply:
>It is a bit deceiving, because yes, while the query is right, the actual active page in the UI doesn't change. 🤔

Actually i tried to clarify that in top section `Batch v/s Datagrid Pagination` that Datagrid pagination is completely different from the concept of Batches.  Do you think adding anything else would help?

Reply:
@PhilippeOberti , implemented your suggestion in fd2ababe4e4